### PR TITLE
PICARD-1796: Date scoring logic

### PR DESCRIPTION
**Pull request by RaysDev**
#  Summary
New logic to include the release date in scoring process.

(Separated out from PR 1771 as you requested)

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other

# Problem
Seems some of the results from AcoustId don't have a release date, and we need a way to de-prioritize those records with lower quality data from the match process.

PICARD-1796

# Solution

Enhanced the scoring process to consider the release date.  It will give highest Part Score to results that have a date that match our own tag, next to records that have a date where year is +/-2 years, next to records that have a non-blank date, next a minimal score if the date is missing, and the lowest Part Score if the date is there but does not match.

# Action
Sorry need little guidance on creating the JIRA you asked to have associated with this.

Please ignore the _config.yml for theme.  Very bizzare that showed up when I first forked, and I definitely did not commit it???

### CODE REVIEW

File: picard/metadata.py [hunk 1]
```diff
@@ -199,6 +200,27 @@ def compare_to_release(self, release, weights):
...
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{4}[-/]", dt):  # 2020-12-31 or 2020/12/31
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{2}/[0-9]{2}/[0-9]{4}", dt):  # 12/31/2020
+            year = dt[6:10]
+        else:
+            year = "0"
+
+        return int(year)
+
```
Comment (zas) [inline picard/metadata.py]: We can do something like:

```python
from collections import Mapping
import re

YEAR_REGEX_YMD = re.compile(r'^(?P<year>\d{4})[-/]?(?:\d{2}[-/]?)?(?:\d{2}[-/]?)?$')
YEAR_REGEX_DMY = re.compile(r'^(?:\d{2}[-/])?(?:\d{2}[-/])?(?P<year>\d{4})$')


def extract_year_from_date(dt):
    """ Tries to handle various date formats, not exhaustive but common, and extract the year"""

    if isinstance(dt, Mapping):
        dt = dt.get('year', '')

    match = YEAR_REGEX_DMY.match(dt)
    if match:
        return int(match.group('year'))

    match = YEAR_REGEX_YMD.match(dt)
    if match:
        return int(match.group('year'))

    return 0


cases = (
    '20201231',
    '202031',
    '20203',
    '2020',
    '2020-12-31',
    '2020/12/31',
    '12/31/2020',
    '12-31-2020',
    '12-2020',
    '01-01-2020',
    'x',
    {'year': '2020'},
    {'year': 'x'},
    {'year': '2020/12/31'}
)

for dt in cases:
    print("%s -> %d" % (dt, extract_year_from_date(dt)))
```


Output:

```
20201231 -> 2020
202031 -> 2020
20203 -> 0
2020 -> 2020
2020-12-31 -> 2020
2020/12/31 -> 2020
12/31/2020 -> 2020
12-31-2020 -> 2020
12-2020 -> 2020
01-01-2020 -> 2020
x -> 0
{'year': '2020'} -> 2020
{'year': 'x'} -> 0
{'year': '2020/12/31'} -> 2020
```


Also, for such function, it is mandatory to write test code (in `test/`).
Also, this function should go in `picard/util` as it's a static method, not specific to `Metadata`.

File: _config.yml [hunk 2]
```diff
@@ -0,0 +1 @@
+theme: jekyll-theme-midnight
```
Comment (zas) [inline _config.yml]: What's this file?

File: picard/file.py [hunk 3]
```diff
@@ -229,7 +230,7 @@ def _preserve_times(self, filename, func):
             # https://docs.python.org/3/library/os.html#os.utime
             # Since Python 3.3, ns parameter is available
             # The best way to preserve exact times is to use the st_atime_ns and st_mtime_ns
-            # fields from the os.stat() result object with the ns parameter to utime.
+            #  fields from the os.stat() result object with the ns parameter to utime.
```
Comment (zas) [inline picard/file.py]: This change is unrelated.

File: picard/file.py [hunk 4]
```diff
@@ -458,7 +459,7 @@ def _move_additional_files(self, old_filename, new_filename):
         new_path = os.path.dirname(new_filename)
         old_path = os.path.dirname(old_filename)
         if new_path == old_path:
-            # skip, same directory, nothing to move
+            #  skip, same directory, nothing to move
```
Comment (zas) [inline picard/file.py]: Unrelated change.

File: picard/metadata.py [hunk 5]
```diff
@@ -218,6 +240,34 @@ def compare_to_release_parts(self, release, weights):
...
+        # release has a date but it does not match ours(all else equal, its better to have an unknown date than a wrong date. since the unknown could actually be correct)
+        factor = 0.0
+        if "date" in release:
+            release_date = release['date']
+            release_year = Metadata.extract_year_from_date(release_date)
+            if "date" in self:
+                metadata_date = self['date']
+                metadata_year = Metadata.extract_year_from_date(metadata_date)
+                if release_date == metadata_date:
+                    factor = 1.00
```
Comment (zas) [inline picard/metadata.py]: You don't need to call `extract_year_from_date()` twice in this case

File: picard/metadata.py [hunk 6]
```diff
@@ -199,6 +200,27 @@ def compare_to_release(self, release, weights):
         sim = linear_combination_of_weights(parts) * get_score(release)
         return SimMatchRelease(similarity=sim, release=release)
 
+    @staticmethod
+    def extract_year_from_date(dt):
```
Comment (samj1912) [inline picard/metadata.py]: You can just use dateutil parser and then get the year.

Comment (zas) [reply picard/metadata.py]: Oh, this too ;)

Comment (samj1912) [reply picard/metadata.py]: https://dateutil.readthedocs.io/en/stable/parser.html#module-dateutil.parser

I would much rather have this (which is thoroughly tested and supported) than trying to implement a bunch of regexes.

File: picard/metadata.py (hunk 1 above)
Comment (phw) [reply picard/metadata.py]: Agreed, especially putting this to util. Also some existing code in Picard assumes dates have the format `yyyy-mm-dd` and year can be extracted as the first four characters. Having this logic in utils allows us to review the cases and reuse the year extraction.

File: _config.yml (hunk 2 above)
Comment (RaysDev) [reply _config.yml]: Let me share in case you run into this again.  I was really stumped on this because I would remember creating a file and committing it, but I didn't.  Apparently while looking for Dark Mode on github.com, I ran into theme selector for "GitHub Pages".   There are icons of various themes, and a button that says "Select Theme."  This button actually creates the _config.yml file in your repo and commits it, without such indication.  Anyway I'll remove the file, but I thought it might save you some trouble in future.

File: picard/file.py (hunk 3 above)
Comment (RaysDev) [reply picard/file.py]: ugh, I turned on some of the auto formatting in my editor, and it got a little over zealous.  Same for next few.

File: picard/metadata.py (hunk 6 above)
Comment (RaysDev) [reply picard/metadata.py]: Done, but I'm surprised we prefer to add a dependency on a third party library for such a small use.

Comment (zas) [reply picard/metadata.py]: > Done, but I'm surprised we prefer to add a dependency on a third party library for such a small use.

It was a suggestion, and I see nothing wrong in using regexes (regexes engines are well-tested too).
But Python has tons of good modules available, and it is good practice to re-use existing stuff whenever possible, but I'm reluctant to rely on third-party libs for another reason: **can anyone there confirm no licensing conflict?**

Comment (phw) [reply picard/metadata.py]: > **can anyone there confirm no licensing conflict?**

We can use this under the https://opensource.org/licenses/BSD-3-Clause, which we can use in our GPL code.

Comment (phw) [reply picard/metadata.py]: I must say I'm also a bit concerned about adding another dependency just for this, so far we had managed to keep the external dependencies to a minimum for Picard. On the other hand we really should review other uses of date handling in our code, I think Picard currently relies on the assumption date is always `YYYY-MM-DD` or any shorter form of this (without day or without month and day). But this might not be true. So having a stable library for date handling might make sense. Two things though:

- `extract_year_from_date` probably should be moved to utils
- `dateutil` must be added to requirements*.txt

**Comment by RaysDev** [comment]
ok, I moved extract_year_from_date() to utils and updated requirements.txt

File: picard/file.py [hunk 7]
```diff
@@ -258,7 +260,7 @@ def _preserve_times(self, filename, func):
             # https://docs.python.org/3/library/os.html#os.utime
             # Since Python 3.3, ns parameter is available
             # The best way to preserve exact times is to use the st_atime_ns and st_mtime_ns
-            # fields from the os.stat() result object with the ns parameter to utime.
+            # fields from the os.stat() result object with the ns parameter to utime.
```
Comment (phw) [inline picard/file.py]: If someone else is wondering: This is changing from a non-breaking space to a regular space. It's strictly speaking an unrelated change, but I don't mind including this minor cleanup in a commit like this. But if someone has a strong opinion on this we could move it to a separate commit.

**Comment by phw** [comment]
@samj1912 @zas This is looking good to me, but I would like to get your final review on this. Especially your opinion on the added dependency.

File: picard/metadata.py [hunk 8]
```diff
@@ -244,6 +247,35 @@ def compare_to_release_parts(self, release, weights):
         except (ValueError, KeyError):
             pass
 
+        # Date Logic - in order of preference
```
Comment (samj1912) [inline picard/metadata.py]: It would be nice to move these comments near the actual else cases as it allows us to understand what's happening.

File: picard/metadata.py [hunk 9]
```diff
@@ -244,6 +247,35 @@ def compare_to_release_parts(self, release, weights):
...
+            release_date = release['date']
+            release_year = extract_year_from_date(release_date)
+            if "date" in self:
+                metadata_date = self['date']
+                if release_date == metadata_date:
+                    factor = 1.00
+                else:
+                    metadata_year = extract_year_from_date(metadata_date)
+                    if release_year == metadata_year:
+                        factor = 0.95
```
Comment (samj1912) [inline picard/metadata.py]: Brownie points if you make these constants with understandable names!

File: picard/util/__init__.py [hunk 10]
```diff
@@ -645,3 +649,15 @@ def limited_join(a_list, limit, join_string='+', middle_string='…'):
     start = a_list[:half]
     end = a_list[-half:]
     return join_string.join(start + [middle_string] + end)
+
+
+def extract_year_from_date(dt):
+    """ Extracts year from  passed in date either dict or string """
+
+    if isinstance(dt, Mapping):
```
Comment (samj1912) [inline picard/util/__init__.py]: I would check for str here instead of mapping. Then return the parsed year. Makes the intention clear.

File: picard/metadata.py (hunk 6 above)
Comment (phw) [reply picard/metadata.py]: @RaysDev  For the requirements we actually need to update not only the requirements.txt but also requirements-win.txt and requirements-macos.txt. We use the later ones to freeze the dependencies for those operating systems as we often encounter issues with specific versions. But for general development we use the requirements.txt as we try to go with newer versions whenever possible.

**Comment by zas** [comment]
> @samj1912 @zas This is looking good to me, but I would like to get your final review on this. Especially your opinion on the added dependency.

Well, I'm fine with extra dependency, if we find out it causes any problem we'll just find an alternative.
For the rest, it looks good to me, but I'd like few minor comments be addressed before merging.

**Comment by RaysDev** [comment]
> For the rest, it looks good to me, but I'd like few minor comments be addressed before merging.

Sorry, I'm not clear at this point which items you would like addressed.  is it the moving/spreading of the comments in metadata.py line 250 "# Date Logic - in order of preference"   ?

**Comment by zas** [comment]
> > For the rest, it looks good to me, but I'd like few minor comments be addressed before merging.
> 
> @zas Sorry, I'm not clear at this point which items you would like addressed. is it the moving/spreading of the comments in metadata.py line 250 "# Date Logic - in order of preference" ?

https://github.com/metabrainz/picard/pull/1509#discussion_r394206521
https://github.com/metabrainz/picard/pull/1509#discussion_r394206258
https://github.com/metabrainz/picard/pull/1509/files#r385592761

File: picard/metadata.py (hunk 9 above)
Comment (RaysDev) [reply picard/metadata.py]: done

File: picard/metadata.py (hunk 8 above)
Comment (RaysDev) [reply picard/metadata.py]: done

File: picard/util/__init__.py (hunk 10 above)
Comment (RaysDev) [reply picard/util/__init__.py]: In previous review above I was asked to change it to check for Mapping.

**Comment by RaysDev** [comment]
All requested changes are completed.

**Comment by zas** [comment]
> All requested changes are completed.

Can you please just squash last two commits (b9868b6 f962c46) in one? It'd be cleaner.

File: picard/metadata.py [hunk 11]
```diff
@@ -153,6 +156,15 @@ class Metadata(MutableMapping):
         ('totaldiscs', 4),
     ]
 
+    __date_match_factors = dict([
```
Comment (samj1912) [inline picard/metadata.py]: Please use normal dict declaration - 

```
VALUES = {
    "key": "value",
    ...
}

**Comment by RaysDev** [comment]
squash completed as requested

File: picard/metadata.py (hunk 11 above)
Comment (RaysDev) [reply picard/metadata.py]: I made the change as requested, and I have no concern with this,  but what I don't understand, is that just one line above, the pre-existing syntax used for __weights uses the syntax I had.  so now just one line apart we have two ways of doing the same thing.  I would think consistency would be more important here.

Comment (zas) [reply picard/metadata.py]: > the pre-existing syntax used for __weights

`__weights` is a list while `__date_match_factors` is a dict.

That said, the syntax you used is perfectly acceptable, but you create a dictionnary by passing a list of tuples to constructor. Using dict literal syntax `{}` is known as being faster (even though the time difference was significantly reduced with Python3 and even more with recent 3.x versions).

See: https://stackoverflow.com/questions/6610606/is-there-a-difference-between-using-a-dict-literal-and-a-dict-constructor

File: picard/metadata.py [hunk 12]
```diff
@@ -244,6 +256,35 @@ def compare_to_release_parts(self, release, weights):
...
+                    elif abs(release_year - metadata_year) <= 2:
+                        date_match_factor = self.__date_match_factors['close_year']
+                    # release has a date but it does not match ours(all else equal, its better to have an unknown date than a wrong date. since the unknown could actually be correct)
+                    else:
+                        date_match_factor = self.__date_match_factors['differed']
+            # release has a date but we don't have one (all else equal, we prefer tracks that have non-blank date values)
+            else:
+                date_match_factor = self.__date_match_factors['exists_vs_null']
+        # release has a no date (all else equal, we don't prefer this release since its date is missing)
+        else:
```
Comment (zas) [inline picard/metadata.py]: I just noticed you placed comments outside blocks they belong to.

I think  it is much better to keep the comment in the code block (and wrap long comments):

```python
       # Date Logic
        date_match_factor = 0.0
        if "date" in release and release['date'] != '':
            release_date = release['date']
            release_year = extract_year_from_date(release_date)
            if "date" in self:
                metadata_date = self['date']
                if release_date == metadata_date:
                    # release has a date and it matches what our metadata had exactly.
                    date_match_factor = self.__date_match_factors['exact']
                else:
                    metadata_year = extract_year_from_date(metadata_date)
                    if release_year == metadata_year:
                        # release has a date and it matches what our metadata had for year exactly.
                        date_match_factor = self.__date_match_factors['year']
                    elif abs(release_year - metadata_year) <= 2:
                        # release has a date and it matches what our metadata had closely (year +/- 2).
                        date_match_factor = self.__date_match_factors['close_year']
                    else:
                        # release has a date but it does not match ours (all else equal,
                        # its better to have an unknown date than a wrong date, since
                        # the unknown could actually be correct)
                        date_match_factor = self.__date_match_factors['differed']
            else:
                # release has a date but we don't have one (all else equal, we prefer
                # tracks that have non-blank date values)
                date_match_factor = self.__date_match_factors['exists_vs_null']
        else:
            # release has a no date (all else equal, we don't prefer this
            # release since its date is missing)
            date_match_factor = self.__date_match_factors['no_release_date']
```

@samj1912 what do you think?

Comment (RaysDev) [reply picard/metadata.py]: done and committed.  thanks for typing it out to show me what you meant.

File: picard/metadata.py (hunk 11 above)
Comment (RaysDev) [reply picard/metadata.py]: I can see the new syntax is better, I just wasn't getting the difference from __weights.  but in tracing through the code, I can see weights is used differently in this class.  Even differently than the `comparison_weights` are used in file.py. which actually is using the same syntax you are suggesting here :)

File: picard/metadata.py (hunk 12 above)
Comment (zas) [reply picard/metadata.py]: Thanks.

Perhaps you can just squash last 4 commits

Comment (RaysDev) [reply picard/metadata.py]: sure, I did the squash

**Comment by phw** [comment]
I added PICARD-1796 to track this. To me this looks good, recent changes improved readability.
